os/kernel/binary_manager/binary_manager_load.c

```diff
@@ -367,19 +364,59 @@ static int loading_thread(int argc, char *argv[])
  * Name: binary_manager_loading
  *
  * Description:
- *   This function create loading thread to load/unload binary.
+ *   This function creates a loading data and add it to loading data list.
+ *   A loading data would be processed by loading thread.
  *
  ****************************************************************************/
-int binary_manager_loading(char *loading_data[])
+int binary_manager_loading(int type, char *data)
 {
 	int ret;
+	loading_data_t *loading_data;
 
-	ret = kernel_thread(LOADINGTHD_NAME, LOADINGTHD_PRIORITY, LOADINGTHD_STACKSIZE, loading_thread, (char * const *)loading_data);
-	if (ret > 0) {
-		bmvdbg("Execute loading thread with pid %d\n", ret);
-	} else {
-		bmdbg("Loading Fail\n");
+	if (type < 0 || type >= LOADCMD_LOAD_MAX) {
+		bmdbg("Invalid loading type %d\n", type);
+		return BINMGR_INVALID_PARAM;
 	}
 
-	return ret;
+	/* Allocate loading data and set values */
+	loading_data = (loading_data_t *)kmm_zalloc(sizeof(loading_data_t));
+	if (loading_data == NULL) {
+		bmdbg("Invalid loading type %d\n", type);
+		return BINMGR_OUT_OF_MEMORY;
+	}
+	loading_data->type = type;
+	if (data != NULL) {
+		strncpy(loading_data->name, data, BIN_NAME_MAX);
+	}
+
+	/* Add loading data to list */
+	sq_addlast((dq_entry_t *)loading_data, &g_loading_list.list);
+	sem_post(&g_loading_list.sem);
+
+	return OK;
+}
+
+/****************************************************************************
+ * Name: binary_manager_initialize_loading
+ *
+ * Description:
+ *   This function create loading thread to load/unload binary
+ *   and initialize resources used in loading thread.
+ *
+ ****************************************************************************/
+int binary_manager_initialize_loading(void)
+{
+	int ret;
+
+	ret = kernel_thread(LOADINGTHD_NAME, LOADINGTHD_PRIORITY, LOADINGTHD_STACKSIZE, loading_thread, NULL);
+	if (ret <= 0) {
+		bmdbg("Failed to create loading thread, errno %d\n", errno);
+		return ERROR;
+	}
+
+	/* Initialize a semaphore and a list of loading data used in loading thread */
+	sem_init(&g_loading_list.sem, 0, 0);
```

Comment:
I think, you'd better initialize before creating loading_thread.

Reply:
I think we don't need to initialize them if creating thread is failed.